File: srcs/juloo.keyboard2/UserDictionaryObserver.java
package juloo.keyboard2;

import android.content.Context;
import android.content.SharedPreferences;
import android.database.ContentObserver;
import android.net.Uri;
import android.os.Handler;
import android.os.Looper;
import android.provider.UserDictionary;
import android.util.Log;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.HashSet;

/**
 * Observes changes to UserDictionary and custom words, providing incremental updates.
 *
 * OPTIMIZATION: Eliminates need for periodic reload checks
 * - ContentObserver detects UserDictionary changes immediately
 * - SharedPreferences listener detects custom word changes
 * - Caches data to avoid repeated content provider queries and JSON parsing
 * - Provides incremental word sets (added/removed) for efficient updates
 *
 * Usage:
 * <pre>
 * UserDictionaryObserver observer = new UserDictionaryObserver(context);
 * observer.setChangeListener(new UserDictionaryObserver.ChangeListener() {
 *   @Override
 *   public void onUserDictionaryChanged(Set<String> addedWords, Set<String> removedWords) {
 *     // Update predictor incrementally
 *   }
 *
 *   @Override
 *   public void onCustomWordsChanged(Map<String, Integer> addedWords,
 *                                    Set<String> removedWords) {
 *     // Update predictor incrementally
 *   }
 * });
 * observer.start();
 *
 * // When done:
 * observer.stop();
 * </pre>
 */
public class UserDictionaryObserver extends ContentObserver
{
  private static final String TAG = "UserDictionaryObserver";

  private final Context _context;
  private final Handler _handler;

  // Cached user dictionary words (for detecting changes)
  private final Map<String, Integer> _cachedUserWords = new HashMap<>();

  // Cached custom words from SharedPreferences
  private final Map<String, Integer> _cachedCustomWords = new HashMap<>();

  // SharedPreferences listener
  private final SharedPreferences.OnSharedPreferenceChangeListener _prefsListener;

  // Change notification listener
  private ChangeListener _changeListener;

  /**
   * Listener interface for dictionary change events.
   */
  public interface ChangeListener
  {
    /**
     * Called when UserDictionary words are added or removed.
     *
     * @param addedWords Words added to UserDictionary (word -> frequency)
     * @param removedWords Words removed from UserDictionary
     */
    void onUserDictionaryChanged(Map<String, Integer> addedWords, Set<String> removedWords);

    /**
     * Called when custom words are added, removed, or modified.
     *
     * @param addedOrModified Words added or with frequency changed (word -> frequency)
     * @param removed Words removed from custom dictionary
     */
    void onCustomWordsChanged(Map<String, Integer> addedOrModified, Set<String> removed);
  }

  public UserDictionaryObserver(Context context)
  {
    super(new Handler(Looper.getMainLooper()));
    _context = context;
    _handler = new Handler(Looper.getMainLooper());

    // Create SharedPreferences listener
    _prefsListener = (prefs, key) -> {
      if ("custom_words".equals(key))
      {
        Log.d(TAG, "Custom words changed in SharedPreferences");
        checkCustomWordsChanges();
      }
    };
  }

  /**
   * Set the change listener for dictionary updates.
   */
  public void setChangeListener(ChangeListener listener)
  {
    _changeListener = listener;
  }

  /**
   * Start observing dictionary changes.
   * Registers ContentObserver and SharedPreferences listener.
   */
  public void start()
  {
    // Register ContentObserver for UserDictionary
    _context.getContentResolver().registerContentObserver(
      UserDictionary.Words.CONTENT_URI,
      true,
      this
    );

    // Register SharedPreferences listener
    SharedPreferences prefs = DirectBootAwarePreferences.get_shared_preferences(_context);
    prefs.registerOnSharedPreferenceChangeListener(_prefsListener);

    // Initial load of cached data
    loadUserDictionaryCache();
    loadCustomWordsCache();

    Log.d(TAG, "Started observing dictionary changes");
  }

  /**
   * Stop observing dictionary changes.
   * Unregisters all observers and listeners.
   */
  public void stop()
  {
    _context.getContentResolver().unregisterContentObserver(this);

    SharedPreferences prefs = DirectBootAwarePreferences.get_shared_preferences(_context);
    prefs.unregisterOnSharedPreferenceChangeListener(_prefsListener);

    Log.d(TAG, "Stopped observing dictionary changes");
  }

  /**
   * Called when UserDictionary content changes.
   */
  @Override
  public void onChange(boolean selfChange, Uri uri)
  {
    Log.d(TAG, "UserDictionary changed: " + uri);
    checkUserDictionaryChanges();
  }

  /**
   * Load UserDictionary into cache.
   */
  private void loadUserDictionaryCache()
  {
    _cachedUserWords.clear();

    try
    {
      android.database.Cursor cursor = _context.getContentResolver().query(
        UserDictionary.Words.CONTENT_URI,
        new String[]{
          UserDictionary.Words.WORD,
          UserDictionary.Words.FREQUENCY
        },
        null,
        null,
        null
      );

      if (cursor != null)
      {
        int wordIndex = cursor.getColumnIndex(UserDictionary.Words.WORD);
        int freqIndex = cursor.getColumnIndex(UserDictionary.Words.FREQUENCY);

        while (cursor.moveToNext())
        {
          String word = cursor.getString(wordIndex).toLowerCase();
          int frequency = (freqIndex >= 0) ? cursor.getInt(freqIndex) : 1000;
          _cachedUserWords.put(word, frequency);
        }

        cursor.close();
        Log.d(TAG, "Loaded " + _cachedUserWords.size() + " user dictionary words into cache");
      }
    }
    catch (Exception e)
    {
      Log.e(TAG, "Failed to load user dictionary cache", e);
    }
  }

  /**
   * Load custom words from SharedPreferences into cache.
   */
  private void loadCustomWordsCache()
  {
    _cachedCustomWords.clear();

    try
    {
      SharedPreferences prefs = DirectBootAwarePreferences.get_shared_preferences(_context);
      String customWordsJson = prefs.getString("custom_words", "{}");

      if (!customWordsJson.equals("{}"))
      {
        org.json.JSONObject jsonObj = new org.json.JSONObject(customWordsJson);
        java.util.Iterator<String> keys = jsonObj.keys();

        while (keys.hasNext())
        {
          String word = keys.next().toLowerCase();
          int frequency = jsonObj.optInt(word, 1000);
          _cachedCustomWords.put(word, frequency);
        }

        Log.d(TAG, "Loaded " + _cachedCustomWords.size() + " custom words into cache");
      }
    }
    catch (Exception e)
    {
      Log.e(TAG, "Failed to load custom words cache", e);
    }
  }

  /**
   * Check for UserDictionary changes and notify listener.
   */
  private void checkUserDictionaryChanges()
  {
    try
    {
      Map<String, Integer> currentWords = new HashMap<>();

      android.database.Cursor cursor = _context.getContentResolver().query(
        UserDictionary.Words.CONTENT_URI,
        new String[]{
          UserDictionary.Words.WORD,
          UserDictionary.Words.FREQUENCY
        },
        null,
        null,
        null
      );

      if (cursor != null)
      {
        int wordIndex = cursor.getColumnIndex(UserDictionary.Words.WORD);
        int freqIndex = cursor.getColumnIndex(UserDictionary.Words.FREQUENCY);

        while (cursor.moveToNext())
        {
          String word = cursor.getString(wordIndex).toLowerCase();
          int frequency = (freqIndex >= 0) ? cursor.getInt(freqIndex) : 1000;
          currentWords.put(word, frequency);
        }

        cursor.close();
      }

      // Compute differences
      Map<String, Integer> addedWords = new HashMap<>();
      Set<String> removedWords = new HashSet<>();

      // Find added words
      for (Map.Entry<String, Integer> entry : currentWords.entrySet())
      {
        if (!_cachedUserWords.containsKey(entry.getKey()))
        {
          addedWords.put(entry.getKey(), entry.getValue());
        }
      }

      // Find removed words
      for (String word : _cachedUserWords.keySet())
      {
        if (!currentWords.containsKey(word))
        {
          removedWords.add(word);
        }
      }

      // Update cache
      _cachedUserWords.clear();
      _cachedUserWords.putAll(currentWords);

      // Notify listener if there are changes
      if (!addedWords.isEmpty() || !removedWords.isEmpty())
      {
        Log.i(TAG, String.format("UserDictionary changed: +%d words, -%d words",
          addedWords.size(), removedWords.size()));

        if (_changeListener != null)
        {
          _changeListener.onUserDictionaryChanged(addedWords, removedWords);
        }
      }
    }
    catch (Exception e)
    {
      Log.e(TAG, "Failed to check user dictionary changes", e);
    }
  }

  /**
   * Check for custom words changes and notify listener.
   */
  private void checkCustomWordsChanges()
  {
    try
    {
      Map<String, Integer> currentWords = new HashMap<>();

      SharedPreferences prefs = DirectBootAwarePreferences.get_shared_preferences(_context);
      String customWordsJson = prefs.getString("custom_words", "{}");

      if (!customWordsJson.equals("{}"))
      {
        org.json.JSONObject jsonObj = new org.json.JSONObject(customWordsJson);
        java.util.Iterator<String> keys = jsonObj.keys();

        while (keys.hasNext())
        {
          String word = keys.next().toLowerCase();
          int frequency = jsonObj.optInt(word, 1000);
          currentWords.put(word, frequency);
        }
      }

      // Compute differences
      Map<String, Integer> addedOrModified = new HashMap<>();
      Set<String> removed = new HashSet<>();

      // Find added or modified words
      for (Map.Entry<String, Integer> entry : currentWords.entrySet())
      {
        Integer cachedFreq = _cachedCustomWords.get(entry.getKey());
        if (cachedFreq == null || !cachedFreq.equals(entry.getValue()))
        {
          addedOrModified.put(entry.getKey(), entry.getValue());
        }
      }

      // Find removed words
      for (String word : _cachedCustomWords.keySet())
      {
        if (!currentWords.containsKey(word))
        {
          removed.add(word);
        }
      }

      // Update cache
      _cachedCustomWords.clear();
      _cachedCustomWords.putAll(currentWords);

      // Notify listener if there are changes
      if (!addedOrModified.isEmpty() || !removed.isEmpty())
      {
        Log.i(TAG, String.format("Custom words changed: +%d words, -%d words",
          addedOrModified.size(), removed.size()));

        if (_changeListener != null)
        {
          _changeListener.onCustomWordsChanged(addedOrModified, removed);
        }
      }
    }
    catch (Exception e)
    {
      Log.e(TAG, "Failed to check custom words changes", e);
    }
  }

  /**
   * Get current cached user dictionary words.
   */
  public Map<String, Integer> getCachedUserWords()
  {
    return new HashMap<>(_cachedUserWords);
  }

  /**
   * Get current cached custom words.
   */
  public Map<String, Integer> getCachedCustomWords()
  {
    return new HashMap<>(_cachedCustomWords);
  }
}

File: srcs/juloo.keyboard2/UserDictionaryObserver.kt
package juloo.keyboard2

import android.content.Context
import android.content.SharedPreferences
import android.database.ContentObserver
import android.database.Cursor
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.provider.UserDictionary
import android.util.Log
import org.json.JSONObject

/**
 * Observes changes to UserDictionary and custom words, providing incremental updates.
 *
 * OPTIMIZATION: Eliminates need for periodic reload checks
 * - ContentObserver detects UserDictionary changes immediately
 * - SharedPreferences listener detects custom word changes
 * - Caches data to avoid repeated content provider queries and JSON parsing
 * - Provides incremental word sets (added/removed) for efficient updates
 *
 * Usage:
 * ```
 * val observer = UserDictionaryObserver(context)
 * observer.setChangeListener(object : UserDictionaryObserver.ChangeListener {
 *   override fun onUserDictionaryChanged(addedWords: Map<String, Int>, removedWords: Set<String>) {
 *     // Update predictor incrementally
 *   }
 *
 *   override fun onCustomWordsChanged(addedWords: Map<String, Int>,
 *                                      removedWords: Set<String>) {
 *     // Update predictor incrementally
 *   }
 * })
 * observer.start()
 *
 * // When done:
 * observer.stop()
 * ```
 */
class UserDictionaryObserver(private val context: Context) : ContentObserver(Handler(Looper.getMainLooper())) {
    companion object {
        private const val TAG = "UserDictionaryObserver"
    }

    private val handler = Handler(Looper.getMainLooper())

    // Cached user dictionary words (for detecting changes)
    private val cachedUserWords: MutableMap<String, Int> = mutableMapOf()

    // Cached custom words from SharedPreferences
    private val cachedCustomWords: MutableMap<String, Int> = mutableMapOf()

    // SharedPreferences listener
    private val prefsListener = SharedPreferences.OnSharedPreferenceChangeListener { _, key ->
        if (key == "custom_words") {
            Log.d(TAG, "Custom words changed in SharedPreferences")
            checkCustomWordsChanges()
        }
    }

    // Change notification listener
    private var changeListener: ChangeListener? = null

    /**
     * Listener interface for dictionary change events.
     */
    interface ChangeListener {
        /**
         * Called when UserDictionary words are added or removed.
         *
         * @param addedWords Words added to UserDictionary (word -> frequency)
         * @param removedWords Words removed from UserDictionary
         */
        fun onUserDictionaryChanged(addedWords: Map<String, Int>, removedWords: Set<String>)

        /**
         * Called when custom words are added, removed, or modified.
         *
         * @param addedOrModified Words added or with frequency changed (word -> frequency)
         * @param removed Words removed from custom dictionary
         */
        fun onCustomWordsChanged(addedOrModified: Map<String, Int>, removed: Set<String>)
    }

    /**
     * Set the change listener for dictionary updates.
     */
    fun setChangeListener(listener: ChangeListener) {
        changeListener = listener
    }

    /**
     * Start observing dictionary changes.
     * Registers ContentObserver and SharedPreferences listener.
     */
    fun start() {
        // Register ContentObserver for UserDictionary
        context.contentResolver.registerContentObserver(
            UserDictionary.Words.CONTENT_URI,
            true,
            this
        )

        // Register SharedPreferences listener
        val prefs = DirectBootAwarePreferences.get_shared_preferences(context)
        prefs.registerOnSharedPreferenceChangeListener(prefsListener)

        // Initial load of cached data
        loadUserDictionaryCache()
        loadCustomWordsCache()

        Log.d(TAG, "Started observing dictionary changes")
    }

    /**
     * Stop observing dictionary changes.
     * Unregisters all observers and listeners.
     */
    fun stop() {
        context.contentResolver.unregisterContentObserver(this)

        val prefs = DirectBootAwarePreferences.get_shared_preferences(context)
        prefs.unregisterOnSharedPreferenceChangeListener(prefsListener)

        Log.d(TAG, "Stopped observing dictionary changes")
    }

    /**
     * Called when UserDictionary content changes.
     */
    override fun onChange(selfChange: Boolean, uri: Uri?) {
        Log.d(TAG, "UserDictionary changed: $uri")
        checkUserDictionaryChanges()
    }

    /**
     * Load UserDictionary into cache.
     */
    private fun loadUserDictionaryCache() {
        cachedUserWords.clear()

        try {
            val cursor: Cursor? = context.contentResolver.query(
                UserDictionary.Words.CONTENT_URI,
                arrayOf(
                    UserDictionary.Words.WORD,
                    UserDictionary.Words.FREQUENCY
                ),
                null,
                null,
                null
            )

            cursor?.use {
                val wordIndex = it.getColumnIndex(UserDictionary.Words.WORD)
                val freqIndex = it.getColumnIndex(UserDictionary.Words.FREQUENCY)

                while (it.moveToNext()) {
                    val word = it.getString(wordIndex).lowercase()
                    val frequency = if (freqIndex >= 0) it.getInt(freqIndex) else 1000
                    cachedUserWords[word] = frequency
                }

                Log.d(TAG, "Loaded ${cachedUserWords.size} user dictionary words into cache")
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to load user dictionary cache", e)
        }
    }

    /**
     * Load custom words from SharedPreferences into cache.
     */
    private fun loadCustomWordsCache() {
        cachedCustomWords.clear()

        try {
            val prefs = DirectBootAwarePreferences.get_shared_preferences(context)
            val customWordsJson = prefs.getString("custom_words", "{}") ?: "{}"

            if (customWordsJson != "{}") {
                val jsonObj = JSONObject(customWordsJson)
                val keys = jsonObj.keys()

                while (keys.hasNext()) {
                    val word = keys.next().lowercase()
                    val frequency = jsonObj.optInt(word, 1000)
                    cachedCustomWords[word] = frequency
                }

                Log.d(TAG, "Loaded ${cachedCustomWords.size} custom words into cache")
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to load custom words cache", e)
        }
    }

    /**
     * Check for UserDictionary changes and notify listener.
     */
    private fun checkUserDictionaryChanges() {
        try {
            val currentWords = mutableMapOf<String, Int>()

            val cursor: Cursor? = context.contentResolver.query(
                UserDictionary.Words.CONTENT_URI,
                arrayOf(
                    UserDictionary.Words.WORD,
                    UserDictionary.Words.FREQUENCY
                ),
                null,
                null,
                null
            )

            cursor?.use {
                val wordIndex = it.getColumnIndex(UserDictionary.Words.WORD)
                val freqIndex = it.getColumnIndex(UserDictionary.Words.FREQUENCY)

                while (it.moveToNext()) {
                    val word = it.getString(wordIndex).lowercase()
                    val frequency = if (freqIndex >= 0) it.getInt(freqIndex) else 1000
                    currentWords[word] = frequency
                }
            }

            // Compute differences
            val addedWords = mutableMapOf<String, Int>()
            val removedWords = mutableSetOf<String>()

            // Find added words
            for ((word, freq) in currentWords) {
                if (!cachedUserWords.containsKey(word)) {
                    addedWords[word] = freq
                }
            }

            // Find removed words
            for (word in cachedUserWords.keys) {
                if (!currentWords.containsKey(word)) {
                    removedWords.add(word)
                }
            }

            // Update cache
            cachedUserWords.clear()
            cachedUserWords.putAll(currentWords)

            // Notify listener if there are changes
            if (addedWords.isNotEmpty() || removedWords.isNotEmpty()) {
                Log.i(TAG, "UserDictionary changed: +${addedWords.size} words, -${removedWords.size} words")

                changeListener?.onUserDictionaryChanged(addedWords, removedWords)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to check user dictionary changes", e)
        }
    }

    /**
     * Check for custom words changes and notify listener.
     */
    private fun checkCustomWordsChanges() {
        try {
            val currentWords = mutableMapOf<String, Int>()

            val prefs = DirectBootAwarePreferences.get_shared_preferences(context)
            val customWordsJson = prefs.getString("custom_words", "{}") ?: "{}"

            if (customWordsJson != "{}") {
                val jsonObj = JSONObject(customWordsJson)
                val keys = jsonObj.keys()

                while (keys.hasNext()) {
                    val word = keys.next().lowercase()
                    val frequency = jsonObj.optInt(word, 1000)
                    currentWords[word] = frequency
                }
            }

            // Compute differences
            val addedOrModified = mutableMapOf<String, Int>()
            val removed = mutableSetOf<String>()

            // Find added or modified words
            for ((word, freq) in currentWords) {
                val cachedFreq = cachedCustomWords[word]
                if (cachedFreq == null || cachedFreq != freq) {
                    addedOrModified[word] = freq
                }
            }

            // Find removed words
            for (word in cachedCustomWords.keys) {
                if (!currentWords.containsKey(word)) {
                    removed.add(word)
                }
            }

            // Update cache
            cachedCustomWords.clear()
            cachedCustomWords.putAll(currentWords)

            // Notify listener if there are changes
            if (addedOrModified.isNotEmpty() || removed.isNotEmpty()) {
                Log.i(TAG, "Custom words changed: +${addedOrModified.size} words, -${removed.size} words")

                changeListener?.onCustomWordsChanged(addedOrModified, removed)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to check custom words changes", e)
        }
    }

    /**
     * Get current cached user dictionary words.
     */
    fun getCachedUserWords(): Map<String, Int> {
        return cachedUserWords.toMap()
    }

    /**
     * Get current cached custom words.
     */
    fun getCachedCustomWords(): Map<String, Int> {
        return cachedCustomWords.toMap()
    }
}
